pkg/ingester/ingester_test.go

```diff
@@ -2960,10 +3015,227 @@ func TestIngester_QueryStreamManySamplesChunks(t *testing.T) {
 
 	// As ingester doesn't guarantee sorting of series, we can get 2 (100k + 500k in first, 1M in second)
 	// or 3 messages (100k or 500k first, 1M second, and 500k or 100k last).
-
 	require.True(t, 2 <= recvMsgs && recvMsgs <= 3)
 	require.Equal(t, 3, series)
-	require.Equal(t, 100000+500000+samplesCount, totalSamples)
+	require.Equal(t, 100000+500000+1000000, totalSamples)
+}
+
+func TestIngester_QueryStream_StreamingWithManySamples(t *testing.T) {
+	// Create ingester.
+	cfg := defaultIngesterTestConfig(t)
+	ctx := user.InjectOrgID(context.Background(), userID)
+
+	c := setupQueryingManySamplesAsChunksTest(ctx, t, cfg)
+
+	s, err := c.QueryStream(ctx, &client.QueryRequest{
+		StartTimestampMs:         0,
+		EndTimestampMs:           1000001,
+		StreamingChunksBatchSize: 64,
+
+		Matchers: []*client.LabelMatcher{{
+			Type:  client.EQUAL,
+			Name:  model.MetricNameLabel,
+			Value: "foo",
+		}},
+	})
+	require.NoError(t, err)
+
+	resp, err := s.Recv()
+	require.NoError(t, err)
+
+	seriesLabelsMsg := client.QueryStreamResponse{
+		Series: []client.QueryStreamSeries{
+			{Labels: mimirpb.FromLabelsToLabelAdapters(labels.FromStrings(labels.MetricName, "foo", "l", "1"))},
+			{Labels: mimirpb.FromLabelsToLabelAdapters(labels.FromStrings(labels.MetricName, "foo", "l", "2"))},
+			{Labels: mimirpb.FromLabelsToLabelAdapters(labels.FromStrings(labels.MetricName, "foo", "l", "3"))},
+		},
+		IsEndOfSeriesStream: true,
+	}
+
+	require.Equal(t, seriesLabelsMsg, *resp)
+
+	recvMsgs := 0
+	series := 0
+	totalSamples := 0
+
+	for {
+		resp, err := s.Recv()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		require.NoError(t, err)
+		require.True(t, len(resp.SeriesChunks) > 0) // No empty messages.
+
+		recvMsgs++
+		series += len(resp.SeriesChunks)
+
+		for _, ts := range resp.SeriesChunks {
+			for _, c := range ts.Chunks {
+				ch, err := chunk.NewForEncoding(chunk.Encoding(c.Encoding))
+				require.NoError(t, err)
+				require.NoError(t, ch.UnmarshalFromBuf(c.Data))
+
+				totalSamples += ch.Len()
+			}
+		}
+	}
+
+	require.Equal(t, 3, recvMsgs) // 1 for each series: second series must be sent in a message of its own
```

Comment:
Are we getting 3 messages here because a series crosses the 1MiB mark? (if yes, worth mentioning here in comments, since batch size is 64 which is not actually hit here)

Reply:
Yep, that's correct. I've clarified the comment in 0ceb3e7.